### PR TITLE
Implement PJR checker

### PR DESCRIPTION
**Pull request by coriolinus**
Continuation of #7191. Adds a function which, given the complete set of voting data and a proposed solution, can determine whether or not that solution satisfies _t_-PJR for some selected _t_.

- [x] implement checker
- [x] ~~determine an appropriate "standard" value for _t_ which we can use at least in the tests~~ develop an intuition for what _t_ means, to aid in test scenario construction
- [x] implement fuzz test which ensures that seq-phragmen without fuzzing always passes the checker

### CODE REVIEW

**Comment by kianenigma** [comment]
> determine an appropriate "standard" value for t which we can use at least in the tests develop an intuition for what t means, to aid in test scenario construction

Also, note that this will probably be a configuration in any case. So we don't need to decide on this now and here.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:65]:
```suggestion
/// subset of the entire NPoS edge graph, encoded as `supports`. This only encodes the
```

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:157]: reference would be nice here.

Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: agree: 4cff663b0

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:58]: what does it means 'balancing parameters'?

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:87]: In this context I would assert in all cases, not just debug.

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:77]: I don't have anything against the approach of using random ids for candidates, but if there are no problems with it, I'd consider using the sorted range as ids, to achieve a simpler code. Example; if you want `x` candidates and `y` votes, candidates will be `0..x` and voters will be `x..x+y`.

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:102]: Again, consider this as a simpler approach https://docs.rs/rand/0.8.3/rand/seq/trait.SliceRandom.html#tymethod.choose

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:57]:
```suggestion
			// Run sesq-phragmen.
```

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:77]: There's a method for this in helpers.rs

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:68]: there's a method for this in the helpers, if you can build a closure that gives the weight.

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs:58]: The `balance` param here: https://github.com/paritytech/substrate/blob/8049fcfa89bfb7375f467a3f22028fea94e7bfe8/primitives/npos-elections/src/phragmen.rs#L66-L71

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs]: Sure: 2451a864d

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs]: Great idea! 1d4e2dba4

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs]: Good idea: fc64ff572

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (kianenigma) [reply primitives/npos-elections/fuzzer/src/common.rs:58]: ahh okay.

**Comment by coriolinus** [comment]
Did some research to determine a small case that fails the fuzzer. This is deterministic: apply the patch in [small_case.diff](https://github.com/paritytech/substrate/files/6034880/small_case.diff.txt), and run the command from the first line of [small_case.txt](https://github.com/paritytech/substrate/files/6034881/small_case.txt), and you'll see roughly the same output. I replaced the candidate and voter IDs with textual identifiers to make life easier, and sorted the voting lists, but that's it for preprocessing. 

In this scenario, there are 12 candidates, 11 winners, and 10 voters. However, none of the voters voted for candidate C or J. Despite that, the algorithm had to pick _someone_ for the 11th winner; it picked C. This appears to me to be a legitimate case where we haven't preserved PJR, because J has a accurate complaint that they had exactly as much voting power allocated to them as C, but didn't make it into the chosen set. 

It's a degenerate case, but then, the point of fuzzing is to discover degenerate cases.

File: primitives/npos-elections/src/pjr.rs
Comment (AlistairStewart) [inline primitives/npos-elections/src/pjr.rs:165]: While the sentence after the colon is correct, the property isn't equal to anything. The test implies the condition, but not every solution that satifies the condition passes the test.

**Comment by AlistairStewart** [comment]
> This appears to me to be a legitimate case where we haven't preserved PJR, because J has a accurate complaint that they had exactly as much voting power allocated to them as C, but didn't make it into the chosen set.

The test should not fail for this. It should ve ok to choose J. Because of rounding errors, it should be ok to choose C even when if J's score is slightly higher than C's support.

Here, the set does not satisfy t-PJR for t=0 but should for any t>0 and the test should show that, because J's score should be 0.

**Comment by coriolinus** [comment]
@AlistairStewart So, the fundamental issue was our unjustified assumption that PJR requires that all elected candidates have non-0 support. 6419d17 removes that requirement. I'll try to find a different failing case now.

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: Updated verbiage in 32da70d.

**Comment by coriolinus** [comment]
I have identified another case in which seq_phragmen produces a result which does not pass PJR. Unfortunately, I couldn't reproduce this case with smaller inputs, which means that this one is moderately large: [failure_medium.txt](https://github.com/paritytech/substrate/files/6042971/failure_medium.txt). (This file used [`symbolicify.py`](https://gist.github.com/coriolinus/d6bfaf490cf1985d99df3cc547fc327c) to generate symbolic representations of the IDs for user friendliness.

In short, the parameters are 87 candidates, 73 winners, 79 voters, and seed 0. The PJR checker says that _t_ is `9 109 055 974 701 609 990`, but Candidate CB has pre-score `11 525 351 653 622 188 319`. 

I don't yet know why `seq_phragmen` produced a non-PJR result here. Running `phragmms` on the same input also produces a result which `pjr_check` rejects.

**Comment by coriolinus** [comment]
Did manage to find a smaller instance of this failure: [fail_medium_2.txt](https://github.com/paritytech/substrate/files/6043466/fail_medium_2.txt)

This one has only 32 candidates, 28 winners, and 5 voters. As before, running the same scenario through `phragmms` also generates a failing winner distribution. This one should be somewhat more tractable to analyze.

**Comment by coriolinus** [comment]
Pulling out the comment from the final commit (as of now), for ease of review:

> It turns out that the fundamental problem causing previous implementations
> of the fuzzer to fail wasn't in `seq_phragmen` _or_ in `pjr_check`: it was
> in the rounding errors introduced in the various conversions between the
> internal data representation and the external one.
> 
> Fixing the fuzzer is then simply an issue of using the internal representation
> and staying in that representation. However, that leaves the issue that
> `seq_phragmen` occasionally produces an output which is technically not
> PJR due to rounding errors. In the future we will need to add some kind of
> "close-enough" threshold. However, that is explicitly out of scope of
> this PR.

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:58]:
```suggestion
const MIN_CANDIDATES: usize = 500;
const MAX_CANDIDATES: usize = 2000;
const MIN_VOTERS: usize = 1000;
const MAX_VOTERS: usize = 5000;
```

These are more realistic polkadot values, and I think it should not get super slow yet.

File: primitives/npos-elections/src/phragmen.rs
Comment (kianenigma) [inline primitives/npos-elections/src/phragmen.rs:67]:
```suggestion
/// Note that rounding errors can potentially cause the output of this function to fail a t-PJR check where t is ....
```

better to specify which PJR test exactly

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:79]:
```suggestion
```

Seems unrelated or outdated.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:103]:
```suggestion
	// collect all candidates and winners into a unified `Vec<CandidatePtr>`.
	let candidates = all_candidates.into_iter().enumerate().map(|(i, c)| {
```

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:130]: I wonder if this can ever happen? No afaik, but we're being defensive here and it is okay.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:163]:
```suggestion
/// `Max {score(c)} < t` where c is every unelected candidate, then this solution is t-PJR.
```

I think this whole sentence is misleading and either need to be reworked or removed. I know that you are trying to address the accuracy issues. But you are phrasing it in a way as if somehow, sometimes, a PJR solution is suddenly no longer PJR. A better assumption is: If a solution `S` is PJR and is no longer PJR due to accuracy issues, it is simply because it is no longer `S` but rather `S'`, a variant thereof.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:195]:
```suggestion
	// First order of business: derive `(candidates, voters)` from `supports`.
```

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:254]:
```suggestion
```

This is no longer relevant.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:265]:
```suggestion
	// NOTE: candidate for saturating_log_sub(). Defensive-only. 
```

Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs:265]: (we always comment such incidents with `defensive` keyword)

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:287]: please use fold and satuaring add.

**Review by kianenigma — APPROVED** [review]
LGTM modulo a few small nits.

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs:58]: Unfortunately, it appears to be pretty slow. On my personal computer, building and executing a single iteration took 159s in release mode. On my Parity Macbook (2019 edition), it took 182s. 

I haven't yet done any real benchmarking to discover whether the slowdown is in case generation or in the `seq_phragmen` algorithm, or both, but I think iterations measured in minutes are too slow for effective fuzzing.

What's clear is that at least one of those two stages is nonlinear in complexity. Cutting your suggested case sizes in half gives a timing of 19.2s on my personal machine, and 22.6s on the MBP. I think that's a better balance of realism and iteration time.

Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs]: 9f928e3a1

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:163]: So, that was in response to https://github.com/paritytech/substrate/pull/8160#discussion_r581943430. I think the actual intent is more like, "There exist solutions which do not pass this test but are nevertheless t-PJR." 

I'll give this some thought and try to clarify.

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: OK: cabce8fd7e7fcb0ae26aa20169ff69c916070a18

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:103]: 06c0e8af8

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: 06c0e8af8

File: primitives/npos-elections/src/phragmen.rs
Comment (coriolinus) [reply primitives/npos-elections/src/phragmen.rs]: Agree: 9c7fb95c7

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: What do you think of the phrasing introduced in eb943752a?

Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs]: Better, but I think it is important to clarify why: accuracy errors.

File: primitives/npos-elections/fuzzer/src/phragmen_pjr.rs
Comment (kianenigma) [reply primitives/npos-elections/fuzzer/src/phragmen_pjr.rs]: Okay no 159 is way too low. Make it something that is feasible on 1 machine. 

I was using the build-host with 128 cores and got a bit spoiled I think :D

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs]: No, I think it's possible for solutions to fail (in edge cases) for reasons that don't derive from accuracy errors. I'm basing that on this text:

> The test implies the condition, but not every solution that satifies the condition passes the test.

File: primitives/npos-elections/src/pjr.rs
Comment (kianenigma) [inline primitives/npos-elections/src/pjr.rs:67]:
```suggestion
	/ committee_size.max(1) as Threshold
```

Overall, always remember that any code that will end up in the wasm runtime must never never ever panic (even in rare/mistake cases like this).

Comment (kianenigma) [reply primitives/npos-elections/src/pjr.rs:67]: (same reason why I asked for sum to change to fold)

File: primitives/npos-elections/fuzzer/src/common.rs
Comment (kianenigma) [inline primitives/npos-elections/fuzzer/src/common.rs:95]: I wasn't convinced about random ids tbh and changed this in my local testing to a sequence. but your final call 😁

Comment (coriolinus) [reply primitives/npos-elections/fuzzer/src/common.rs:95]: I don't think it's that much more expensive to randomize them, and I prefer "more natural" fuzzing inputs when possible.

File: primitives/npos-elections/src/pjr.rs
Comment (coriolinus) [reply primitives/npos-elections/src/pjr.rs:67]: Good catch!

File: primitives/npos-elections/src/lib.rs
Comment (shawntabrizi) [inline primitives/npos-elections/src/lib.rs:342]:
```suggestion
```

File: primitives/npos-elections/src/pjr.rs
Comment (shawntabrizi) [inline primitives/npos-elections/src/pjr.rs:3]:
```suggestion
// Copyright (C) 2021 Parity Technologies (UK) Ltd.
```

**Review by shawntabrizi — APPROVED** [review]
looks very clean to me

**Comment by coriolinus** [comment]
bot merge

**Pull request merged**
